src/cargo/util/config/mod.rs

```diff
@@ -1244,6 +1248,39 @@ impl Config {
         &self.progress_config
     }
 
+    /// Create an EnvConfigValue hashmap from the "env" table
+    fn get_env_config(&self) -> CargoResult<EnvConfig> {
```

Comment:
Would it be possible to do custom serde deserialization (e.g. via `#[serde(untagged)]`) instead of having custom deserialization logic here?

Reply:
So I tried that originally and could not get it to work. I had something like:
```
#[derive(Debug, Deserialize)]
#[serde(transparent)]
pub struct EnvConfig {
    vars: HashMap<String, EnvConfigValue>
}

#[derive(Debug, Deserialize)]
#[serde(untagged)]
pub enum EnvConfigValue {
    Simple(String),
    WithOptions{
        value: Value<String>,
        force: bool,
        relative: bool, 
    }
}

```

The problem was serde would never parse the `Value<T>` type in the `WithOptions` variant, it worked if it was a regular `String`. I'm not sure why, maybe the entire "path" from the root of the deserialization needs to be `Value<T>` types in order for them to be handle correctly in some inner struct?

Reply:
I think this structure may work instead?

```rust
#[derive(Debug, Deserialize)]
#[serde(untagged)]
pub enum EnvConfigValue {
    Simple(String),
    WithOptions(WithOptions),
}

#[derive(Debug, Deserialize)]
pub struct WithOptions {
        value: Value<String>,
        force: bool,
        relative: bool, 
}
```

Reply:
No, that doesn't work either. It seems like `untagged` and the `Value<T>` types do not play nicely together, it will never select a variant with that type. Latest commit does contain a struct/enum layout that does work though. So no more custom logic.